### ISSUE TITLE
Cross-tenant queries produce cardinality estimation cache keys that are too long

### ISSUE THREAD

**Issue by chencs** [grafana/mimir#6931]
#### Describe the bug
When running cross-tenant queries with the query-frontend results cache enabled, the query caching middleware attempts to concatenate _all_ tenant IDs into the key ([code](https://github.com/grafana/mimir/blob/efa26e743027bf495c340a893939b2671ba05959/pkg/frontend/querymiddleware/split_and_cache.go#L173)). There's a 250-byte limit on memcached keys ([code](https://github.com/grafana/gomemcache/blob/7de47a8c3cb08c073e448e0a066e4f0e1a283928/memcache/memcache.go#L616)), so it's possible to have enough tenants (or long enough tenant names) that the key ends up exceeding that 250-byte limit, at which point gets will fail.

#### To Reproduce

1. Create multiple tenants whose names, concatenated, exceed 250 bytes in length.
2. Query across all tenants using [the wildcard tenant](https://grafana.com/blog/2021/08/26/whats-new-in-grafana-enterprise-metrics-1.5-per-tenant-usage-metrics-and-a-wildcard-tenant-for-queries/).
3. See malformed key logs on memcached fetches.

#### Expected behavior

Keys should be short enough that they don't exceed the 250-byte limit. Perhaps that means that cache keys should be created and looked up on a per-tenant basis?

**Comment by 56quarters** [comment]
One potential solution to this is to sort the tenant IDs (the parsing logic might already do this) and hash them and use the hash in the cache key instead of the tenant IDs directly.

**Comment by dimitarvdimitrov** [comment]
the cache key is then hashed and that's used to key memcached. The full key is stored as part of the item to check for hash collisions, but it's not included verbatim in the cache key.

https://github.com/grafana/mimir/blob/efa26e743027bf495c340a893939b2671ba05959/pkg/frontend/querymiddleware/split_and_cache.go#L458

**Comment by chencs** [comment]
Ah, you're right, and then the hashed key gets read back out [here](https://github.com/grafana/mimir/blob/efa26e743027bf495c340a893939b2671ba05959/pkg/frontend/querymiddleware/split_and_cache.go#L346), I missed that. So this isn't a problem for the results cache, but it seems like it _is_ a problem for _a_ cache somewhere, likely the cardinality estimation cache, since I do see the [`QS` prefix](https://github.com/grafana/mimir/blob/a5b47fb41ef125b812dcbd05d71e36459066022f/pkg/frontend/querymiddleware/cardinality.go#L160-L176) in logs.

**Comment by rnerkar** [comment]
It seems we have been hitting the same issue since adding new tenants 2 days back 
![image](https://github.com/grafana/mimir/assets/17119137/81803bea-f90d-4eb3-b04c-0165a9a6d932)

We have the below in logs for query-frontend

{"time":"2023-12-19T17:27:04.952051693Z","message":"ts=2023-12-19T17:27:04.950734513Z caller=memcached_client.go:335 level=warn name=frontend-cache msg=\"failed to fetch items from memcached\" numKeys=1 firstKey=1@QS:<tenant1>|<tenant2>|...|<tenantX>:819455e299e84fc0:236529:0 err=\"malformed: key is too long or contains invalid characters\"", "tenant":"somename"}

When I check the length of all the tenants, it is exceeding 250 characters
@chencs Are you seeing similar log entries ?

**Comment by chencs** [comment]
@rnerkar yes, except my logs don't include that `tenant` key. My understanding is that the cardinality estimation cache is used to _limit_ the number of shards a query can be broken into for parallelization ([code](https://github.com/grafana/mimir/blob/d8678ef08465d63ff09a96afaa64e66d9c976ddf/pkg/frontend/querymiddleware/querysharding.go#L294)). 

From reading [this issue](https://github.com/grafana/mimir/issues/4108) describing the goals for implementing cardinality estimation for query sharding, I believe this means that the failure mode (not being able to fetch cardinality estimation cache keys) potentially results in unnecessary over-sharding (a nuisance with respect to overhead), but not unnecessary under-sharding (OOMs resulting in query failure).

**Comment by rnerkar** [comment]
Thanks for explaining @chencs 
So with the understanding that there are no adverse effects on Mimir's performance with these errors, except for the fact that small queries will get over-sharded
Do we have a fix coming in, or something to be done from our end?
We have had talks about using redis instead of Memcached to get around this issue but that's still not finalized
Will wait till I hear back from you guys. Thanks in advance !

**Comment by 56quarters** [comment]
> Do we have a fix coming in, or something to be done from our end?

We apply a hash to keys to make them a consistent length elsewhere, we'll do the same here. No ETA but it should be a simple fix.

> We have had talks about using redis instead of Memcached to get around this issue but that's still not finalized
> Will wait till I hear back from you guys. Thanks in advance !

We don't recommend using Redis since we (Grafana) don't use it in production and it's not part of the Helm chart or Jsonnet.

**Comment by nabadger** [comment]
We're hitting this issue as well. 

Note that it also seems to increment `thanos_memcached_operation_failures_total`, so we actually get alerted on this.

**Comment by rishabhkumar92** [comment]
We are hitting into this issue, I saw we close the change that was opened earlier.

**Comment by chencs** [comment]
Implementing the fix for this isn't currently planned, since the proposed quick fix above did not turn out to be as simple as expected. @56quarters did list some alternative options in the related PR before closing, though.

**Comment by chencs** [comment]
That PR was for the error cache (fast-failing requests on non-retryable errors), not cardinality estimation cache keys. This work is still in the backlog but not planned.

**Comment by kennytrytek-wf** [comment]
>...the failure mode (not being able to fetch cardinality estimation cache keys) potentially results in unnecessary over-sharding (a nuisance with respect to overhead)...

I wanted to highlight that over-sharding can cause significant resource usage, and increase cost of ownership by 3x or more. Our setup uses `query_sharding_target_series_per_shard: 2500`, which results in ~4,000 queries per second total to the ingesters in a particular environment. If I remove that setting and use the default of `0`, ingester queries shoot up to ~40,000 queries per second, 10x the volume.

Target series per shard is _critical_ to the proper functioning of Mimir, so this fix should be prioritized for release as soon as reasonably possible. I understand the use-case for cross-tenant queries is a smaller percentage of users, though many of our most important overview dashboards are cross-tenant for 100+ tenants with a key length over 1800 characters. I am hopeful this change will improve query performance on those dashboards.
